### PR TITLE
fix(nvidia/hw-slowdown): rename from "clock" to only expose hardware slowdown issues, convert to events

### PR DESCRIPTION
**Pull request by gyuho**
- Rename `components/clock` to `components/hw-slowdown` to clarify that we only tracks the critical hw slowdown clock events
- No NVML-based clock events in `components/hw-slowdown` `/states`, instead supports `/events` API (similarly to Xid component)
- `components/hw-slowdown` `/states` API still returns nvidia-smi based HW slowdown clock events, as a safety fallback
- Do not expose non-HW slowdown clock events (but `nvml` package flags it if other events happen, in case we want to expose this in the future)
- Using NVML mock to unit test the logic

### CODE REVIEW

**Comment by xiang90** [comment]
this is still way too complicated than what it should be.

split out the hw slowdown checking. only use nvidia smi (why not?!)

**Comment by xiang90** [comment]
rename this whole pkg to hardware slowdown. and ONLY check hardware slowdown by using nvidia smi -q

File: components/accelerator/nvidia/query/nvml/clock_events.go [hunk 1]
```diff
@@ -118,52 +123,111 @@ func GetClockEvents(uuid string, dev device.Device) (ClockEvents, error) {
...
+	hwSlowdown  bool
+}
+
 // ref. https://github.com/NVIDIA/go-nvml/blob/main/gen/nvml/nvml.h
-var clockEventReasons = map[uint64]string{
+// ref. https://docs.nvidia.com/deploy/nvml-api/group__nvmlClocksEventReasons.html
+var clockEventReasonsToInclude = map[uint64]reasonType{
 	// ref. nvmlClocksEventReasonGpuIdle
-	0x0000000000000001: "GPU is idle and clocks are dropping to Idle state",
+	0x0000000000000001: {
```
Comment (xiang90) [inline components/accelerator/nvidia/query/nvml/clock_events.go:181]: define all as const around line 166. not just hw related

File: components/accelerator/nvidia/clock/component_output.go [hunk 2]
```diff
@@ -92,26 +92,17 @@ func ParseStatesToOutput(states ...components.State) (*Output, error) {
...
+			clockEventsHWReasons = append(clockEventsHWReasons, clockEvents.HWSlowdownReasons...)
 		}
 	}
 	if len(o.HWSlowdownSMI.Errors) > 0 {
-		clockEventsReasons = append(clockEventsReasons, o.HWSlowdownSMI.Errors...)
+		clockEventsHWReasons = append(clockEventsHWReasons, o.HWSlowdownSMI.Errors...)
 	}
 
-	if len(clockEventsReasons) == 0 {
+	if len(clockEventsHWReasons) == 0 {
```
Comment (xiang90) [inline components/accelerator/nvidia/clock/component_output.go:105]: need to fix line 110. `no hardware slowdown found`

File: components/accelerator/nvidia/clock/component_output.go [hunk 3]
```diff
@@ -125,7 +116,7 @@ func (o *Output) States() ([]components.State, error) {
 		}, nil
 	}
 
-	yb, err := yaml.Marshal(clockEventsReasons)
+	yb, err := yaml.Marshal(clockEventsHWReasons)
 	if err != nil {
 		return nil, err
 	}
```
Comment (xiang90) [inline components/accelerator/nvidia/clock/component_output.go:122]: need to fix line 127... the current format is not readable 

```
clock events found - 'GPU-16d6b3a3-a908-7304-099f-2141543e20d6: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-2ac2e70a-0faa-6f55-cdf4-035f13f6af7f: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-4ba57800-cb2f-bad7-d56d-f426bf0893b5: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-78a2612f-5bc8-e552-96b1-1385e4356f8c: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-81dc2e79-bac4-2119-d12e-5f5c7857cd2c: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-a00045ab-26c6-09c9-f8bb-006390140ab7: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-bffd88ce-ab19-0f0d-cd0f-c92a04196c1f: GPU is idle and clocks are dropping to Idle state (nvml)' - 'GPU-e19ab0d0-e85a-e978-6345-8958c626da1d: GPU is idle and clocks are dropping to Idle state (nvml)'
```

File: components/accelerator/nvidia/query/nvml/clock_events.go [hunk 4]
```diff
@@ -118,52 +123,140 @@ func GetClockEvents(uuid string, dev device.Device) (ClockEvents, error) {
...
+
 // ref. https://github.com/NVIDIA/go-nvml/blob/main/gen/nvml/nvml.h
-var clockEventReasons = map[uint64]string{
+// ref. https://docs.nvidia.com/deploy/nvml-api/group__nvmlClocksEventReasons.html
+var clockEventReasonsToInclude = map[uint64]reasonType{
 	// ref. nvmlClocksEventReasonGpuIdle
-	0x0000000000000001: "GPU is idle and clocks are dropping to Idle state",
+	reasonGPUIdle: {
+		description: "GPU is idle and clocks are dropping to Idle state",
+		hwSlowdown:  false,
```
Comment (xiang90) [inline components/accelerator/nvidia/query/nvml/clock_events.go:189]: isHWSlowdown.

**Comment by xiang90** [comment]
defer to @cardyok. we need to find a way to test this end to end.

File: components/accelerator/nvidia/hw-slowdown/component.go [hunk 5]
```diff
@@ -39,17 +44,18 @@ type component struct {
...
 
 func (c *component) States(ctx context.Context) ([]components.State, error) {
 	last, err := c.poller.Last()
 	if err == query.ErrNoData { // no data
-		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", Name)
+		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", nvidia_hw_slowdown_id.Name)
 		return []components.State{
 			{
-				Name:    Name,
+				Name:    nvidia_hw_slowdown_id.Name,
```
Comment (cardyok) [inline components/accelerator/nvidia/hw-slowdown/component.go:58]: maybe use StateNameHWSlowdown here to keep state output consistent

File: components/accelerator/nvidia/hw-slowdown/component.go [hunk 6]
```diff
@@ -84,7 +90,7 @@ func (c *component) States(ctx context.Context) ([]components.State, error) {
 		cs := make([]components.State, 0)
 		for _, e := range allOutput.SMIQueryErrors {
 			cs = append(cs, components.State{
-				Name:    Name,
+				Name:    nvidia_hw_slowdown_id.Name,
```
Comment (cardyok) [inline components/accelerator/nvidia/hw-slowdown/component.go:93]: same here

Comment (gyuho) [reply components/accelerator/nvidia/hw-slowdown/component.go]: done

File: components/accelerator/nvidia/query/nvml/clock_events.go [hunk 7]
```diff
@@ -112,58 +125,201 @@ func GetClockEvents(uuid string, dev device.Device) (ClockEvents, error) {
...
+}
+
+func (inst *instance) ClockEventsSupported() bool {
+	inst.mu.RLock()
+	defer inst.mu.RUnlock()
+
+	return inst.clockEventsSupported
+}
+
+func (inst *instance) RecvClockEventsHWSlowdown() <-chan *ClockEvents {
```
Comment (cardyok) [inline components/accelerator/nvidia/query/nvml/clock_events.go:256]: seems currently no one is using this function

Comment (gyuho) [reply components/accelerator/nvidia/query/nvml/clock_events.go]: right, removed

**Comment by cardyok** [comment]
lgtm, manual tested injecting hw slowdown event

**Review by cardyok — APPROVED** [review]

**Pull request merged**